### PR TITLE
[DCP] Add process group aware planner to PT

### PR DESCRIPTION
**Pull request by wz337**
Upstream pg aware planners and its test from Tau:
https://github.com/pytorch/tau/blob/main/spmd/checkpoint/pg_planner.py
https://github.com/pytorch/tau/blob/main/test/spmd/checkpoint/test_pg_planner.py

This enables FSDP+TP using LOCAL_STATE_DICT.

### CODE REVIEW

File: torch/distributed/checkpoint/pg_aware_planner.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    It identifies ShardedTensor instances that are using the non-default
+    process groups and change their FQNs to be prefixed by a PG-specific string.
+    """
+    state_dict_copy = {}
+
+    for k, v in state_dict.items():
+        # Only rename the fqn if the current process group is a sub-process group.
+        if (
+            isinstance(v, ShardedTensor)
+            and v._process_group != dist.distributed_c10d._get_default_group()
```
Comment (kumpera) [inline torch/distributed/checkpoint/pg_aware_planner.py:39]: `dist.distributed_c10d._get_default_group()` -> `dist.group.WORLD`. We want to reduce usage of _get_default_group outside of core c10d.

File: torch/distributed/checkpoint/pg_aware_planner.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+)
+from torch.distributed.checkpoint.planner import ReadItem
+from torch.distributed._shard.sharded_tensor.api import ShardedTensor
+
+
+def get_ranks(pg: c10d.ProcessGroup) -> List[int]:
+    """
+    Return an array of global ranks for a given process group.
+    """
+    return [c10d.get_global_rank(pg, i) for i in range(pg.size())]
```
Comment (kumpera) [inline torch/distributed/checkpoint/pg_aware_planner.py:24]: This function is the same as `dist.get_process_group_ranks`.

File: test/distributed/checkpoint/test_pg_aware_planner.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        """
+
+        with FSDP.state_dict_type(model, StateDictType.LOCAL_STATE_DICT):
+            state_dict = model.state_dict()
+
+            cp.save_state_dict(
+                state_dict=state_dict,
+                storage_writer=cp.FileSystemWriter(path=CHECKPOINT_DIR),
+                planner=ProcessGroupAwareSavePlanner(
+                    flatten_sharded_tensors=False,
```
Comment (kumpera) [inline test/distributed/checkpoint/test_pg_aware_planner.py:121]: Those 3 cause issues if enabled?

File: test/distributed/checkpoint/test_pg_aware_planner.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    def test_process_group_aware_planner(self) -> None:
+        CHECKPOINT_DIR = self.temp_dir
+
+        model = MyModule(
+            rank=dist.get_rank(),
+            extra_state=0,
+            extra_state_tensor=torch.tensor([[1.0, -1.0], [1.0, -1.0]]),
+        ).cuda(dist.get_rank())
+
+        model = FSDP(model, process_group=self._create_new_dist_group())
```
Comment (kumpera) [inline test/distributed/checkpoint/test_pg_aware_planner.py:97]: We should explain that by using 2 PGs we're creating two in-memory models in the cluster whose only difference is the PG they got assigned.

File: torch/distributed/checkpoint/pg_aware_planner.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            state_dict_copy[fqn] = v
+        else:
+            state_dict_copy[k] = v
+
+    return state_dict_copy
+
+
+class ProcessGroupAwareSavePlanner(DefaultSavePlanner):
+    """
+    ProcessGroupAwareSavePlanner extends DefaultSavePlanner and re-write the state_dict.
```
Comment (kumpera) [inline torch/distributed/checkpoint/pg_aware_planner.py:52]: We should update the docstring to explain how this planner can be used.

**Review by kumpera — APPROVED** [review]

File: test/distributed/checkpoint/test_pg_aware_planner.py (hunk 4 above)
Comment (wz337) [reply test/distributed/checkpoint/test_pg_aware_planner.py:97]: Sounds good. Will update in a separate PR.

File: torch/distributed/checkpoint/pg_aware_planner.py (hunk 5 above)
Comment (wz337) [reply torch/distributed/checkpoint/pg_aware_planner.py:52]: Sounds good. Will update in a separate PR.

**Pull request closed without merging**